### DISCUSSION TITLE
Multi-Timbral UI

### DISCUSSION THREAD

**Discussion by winder** [ElectroTechnique/TSynth-Teensy4.1#78] [T4.1 code development]
Now that the code supports multiple timbres we need a way to configure it. I have a lot of ideas but am not quite sure where to start.

The main constraint is that voices must be assigned to a particular timbre, so some upfront configuration is required.

Ideas:
* Assign timbre to MIDI channel(s).
* Last-note-played configuration. This is what I used for the prototype, but not great if you had a sequence playing and you wanted to adjust lead sounds.
* Press + Hold encoder to open a new menu:
    * New menu contains [ `1. <patch>`, `Create...` ], whichever is highlighted is enabled for configuration.
        * Where does `Create...` get voices from? Evenly distribute by default or configurable?
    * If you select one of the timbres you get a menu with [ `Num voices`, `MIDI Channel`, `Rename...`, `Delete...` ]
        * A "split keyboard" would be cool. How to configure that? Requires offsetting the MIDI note?
        * What to do with voices when you increase or decrease the number?
        * Name the timbres?
        * I really like that this method supports layering by default, just assign multiple patches to the same channel.
        * For layering it might be nice to have the note activate after an initial delay
* Rotary encoder configuration
    * Configure the rotary encoder default selection: [ `Patch`, `Timbre`, `Timbre Config` ]
* Configuration files
    * It is probably useful to name the timbres, so that would need to be saved somewhere.
    * It may also be nice to support multiple configurations.

**Comment by Electrotechnique** [comment]
To let you know - pressing and holding the encoder currently reloads the current patch, but it could be used for this configuration. I have a bunch more questions actually. I don't mind if multi-timbrality is limited to two patches if it makes things simpler:

How will users change control between each patch?

A simple file with a name plus the patch names, MIDI channels and any config data (split etc). A problem is when patches are deleted, which is why patch name instead of number would be used?

Yes, splitting the keyboard could be done by pressing the note where the split occurs and then offsetting every note on/off above/below that by a certain amount, but then how is this configurable so that patches aren't played too high or low?

**Reply by winder** [reply]
> To let you know - pressing and holding the encoder currently reloads the current patch, but it could be used for this configuration. 

Good to know, I've been switching one patch up then back down :)

> I have a bunch more questions actually. I don't mind if multi-timbrality is limited to two patches if it makes things simpler:
> 
> How will users change control between each patch?

I'm hoping that someone familiar with other synths has seen this feature before a knows of a good approach used by those instruments.

My main goal is to chain together some MIDI controllers and do some jamming, I'm not concerned with recording / track isolation / DAW support. So I'm thinking of ways to support lots of timbres, plus your idea of layering is really intriguing so that would need many sounds as well.

My current thinking is as follows...

**press + hold recall** => enter a menu like the patch menu, but for the current timbers. The leading number indicates what channel activates the patch, here you can see a note sent on channel 1 activates two patches.
- 1 | lead layer 1
- 1 | lead layer 2
- 2 | bass
- < New Timbre >

**recall with timbre selected** => it becomes the active patch and modifications can be made as usual.
**recall with < New > selected** => enter the settings menu, where the name can be set.
**settings** => enter a settings menu with options:

* Name | When selected opens the name entry routine - perhaps with an improved rename functionality. 
* Patch | This would need to display the patch selection menu. 
* Midi Ch | Select channel 1-16. 
* Num Voices | Select the number of voices. I'm not exactly sure how this one will work, somehow it needs to steal voices from other timbres or indicate that too many voices have been allocated. Maybe marking things red and showing a prompt to fix voice numbers. <br> Alternatively maybe changing the number of voices has a new config screen that lets all voice numbers be adjusted at the same time. 

I'm not sure how different timbre profiles can be loaded yet. Maybe `< New >`, `< New Profile >` and `< Load Profile >`.

> 
> A simple file with a name plus the patch names, MIDI channels and any config data (split etc). A problem is when patches are deleted, which is why patch name instead of number would be used?

I haven't fully considered this yet. I think you're right, the patch name needs to be used. The delete option also needs to be changed so that it verifies that no timbre profiles are using the patch.

> 
> Yes, splitting the keyboard could be done by pressing the note where the split occurs and then offsetting every note on/off above/below that by a certain amount, but then how is this configurable so that patches aren't played too high or low?

Split mode is pretty low priority. It's probably pretty complicated, I think some MIDI keyboards support it natively so maybe this is left as a controller feature for now.

**Comment by Electrotechnique** [comment]
This is how a [Waldorf Blofeld deals with multi-timbrality](https://www.youtube.com/watch?v=dylOTZAaGuA). It accesses it with a dedicated button and then pressing Play and scrolling, cycles through each patch (part). Then there are a load of other options for transposing, detuning, velocity, split settings..... for each one. I don't suggest we get too ambitious though. 

The display on TSynth is colour and comparatively high resolution but quite small and we are limited to one encoder and possibly just the Settings button. If you need a new font or symbols to be made, just ask and I'll sort it out. I think your ideas for this are fine, be careful with using patch name to identify patches as they are obviously not unique. I wonder if each patch needs a UUID type thing?

For number of voices - default by dividing 12 by the number of patches and that's what they're given?

**Reply by winder** [reply]
Ok, I'll start with implementing what I've described above. I think the basic idea without anything extra could already allow for some really interesting sounds.

The nice part about all of the recent refactoring is that most of these changes can be done one at a time without adding too much extra complexity to the codebase. I want to try adding as many cool features as possible, we'll see what makes sense.
